xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,444 @@
+  XdsServerWrapper(
+          String listenerAddress,
+          ServerBuilder<?> delegateBuilder,
+          XdsServingStatusListener listener,
+          AtomicReference<FilterChainSelector> filterChainSelectorRef,
+          XdsClientPoolFactory xdsClientPoolFactory,
+          ScheduledExecutorService timeService) {
+    this.listenerAddress = checkNotNull(listenerAddress, "listenerAddress");
+    this.delegateBuilder = checkNotNull(delegateBuilder, "delegateBuilder");
+    this.listener = checkNotNull(listener, "listener");
+    this.filterChainSelectorRef = checkNotNull(filterChainSelectorRef, "filterChainSelectorRef");
+    this.xdsClientPoolFactory = checkNotNull(xdsClientPoolFactory, "xdsClientPoolFactory");
+    this.timeService = checkNotNull(timeService, "timeService");
+    this.delegate = delegateBuilder.build();
+  }
+
+  @Override
+  public Server start() throws IOException {
+    checkState(started.compareAndSet(false, true), "Already started");
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalStart();
+      }
+    });
+    Exception exception;
+    try {
+      exception = initialStartFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (exception != null) {
+      throw (exception instanceof IOException) ? (IOException) exception :
+              new IOException(exception);
+    }
+    return this;
+  }
+
+  private void internalStart() {
+    try {
+      xdsClientPool = xdsClientPoolFactory.getOrCreate();
+    } catch (Exception e) {
+      StatusException statusException = Status.UNAVAILABLE.withDescription(
+              "Failed to initialize xDS").withCause(e).asException();
+      listener.onNotServing(statusException);
+      initialStartFuture.set(statusException);
+      return;
+    }
+    xdsClient = xdsClientPool.getObject();
+    // TODO(chengyuanzhang): add an API on XdsClient indicating if it is using v3, don't get
+    //  from bootstrap.
+    boolean useProtocolV3 = xdsClient.getBootstrapInfo().getServers().get(0).isUseProtocolV3();
+    String listenerTemplate = xdsClient.getBootstrapInfo().getServerListenerResourceNameTemplate();
+    if (!useProtocolV3 || listenerTemplate == null) {
+      StatusException statusException =
+          Status.UNAVAILABLE.withDescription(
+              "Can only support xDS v3 with listener resource name template").asException();
+      listener.onNotServing(statusException);
+      initialStartFuture.set(statusException);
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+      return;
+    }
+    discoveryState = new DiscoveryState(listenerTemplate.replaceAll("%s", listenerAddress));
+  }
+
+  @Override
+  public Server shutdown() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        if (!delegate.isShutdown()) {
```

Comment:
Based on the recent discussion we had it looks like this check is unnecessary. But my original comment was based on a similar check in other places (including this PR). Also saw this comment https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/InternalServer.java#L46 but then realized it is for `InternalServer` and for the `Server` hierarchy it doesn't apply.

I am okay with this so this is not a blocker but if you want to remove the unnecessary check in all the places I am okay with that too

Reply:
yea it should not be wrong but doing the check is a little optimization so i would just keep it.